Diffing maps with pointer values fails

When diffing a map with pointer values, e.g.:
```
struct1 := tmstruct{Bar: 1, Foo: "one"}
struct2 := tmstruct{Bar: 2, Foo: "two"}

map1 := map[string]*tmstruct{
  "one": &struct1,
}
map2 := map[string]*tmstruct{
"one": &struct1,
"two": &struct2,
}
```
...this comes back as ErrTypeMismatch due to comparing nil to a pointer.

I was able to fix this for my use case by modifying `diff_pointer.go` thusly:
```
func (cl *Changelog) diffPtr(path []string, a, b reflect.Value) error {
	if a.Kind() == reflect.Invalid {
		cl.add(CREATE, path, nil, b.Interface())
		return nil
	}

	if b.Kind() == reflect.Invalid {
		cl.add(DELETE, path, a.Interface(), nil)
		return nil
	}
	[...]
```

However, this change breaks the tests **mismatched-values-struct-nil** and **mismatched-values-nil-struct**.

Do you have thoughts on how this conflict could be resolved?

For testing my change, I added these cases to the case list in TestDiff:
```
		{
			"map-string-pointer-create-test",
			map[string]*tmstruct{"one": &struct1},
			map[string]*tmstruct{"one": &struct1, "two": &struct2},
			Changelog{
				Change{Type: CREATE, Path: []string{"two"}, From: nil, To: &struct2},
			},
			nil,
		},
		{
			"map-string-pointer-delete-test",
			map[string]*tmstruct{"one": &struct1, "two": &struct2},
			map[string]*tmstruct{"one": &struct1},
			Changelog{
				Change{Type: DELETE, Path: []string{"two"}, From: &struct2, To: nil},
			},
			nil,
		},
```

I tried your branch out and ran into a failure when I run it through my test suite, with this panic chain:
```
panic: reflect.Value.Interface: cannot return value obtained from unexported field or method [recovered]
        panic: reflect.Value.Interface: cannot return value obtained from unexported field or method

goroutine 26 [running]:
testing.tRunner.func1(0xc0000e0900)
        /home/erhudy/go1.12.5/src/testing/testing.go:830 +0x392
panic(0xb868c0, 0xd74bc0)
        /home/erhudy/go1.12.5/src/runtime/panic.go:522 +0x1b5
reflect.valueInterface(0xb868c0, 0xc0001c8f20, 0x1b8, 0xc83c01, 0xb, 0x4)
        /home/erhudy/go1.12.5/src/reflect/value.go:990 +0x1bf
reflect.Value.Interface(...)
        /home/erhudy/go1.12.5/src/reflect/value.go:979
bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff.(*Changelog).diffString(0xc00012ea60, 0xc0001c8fc0, 0x3, 0x3, 0xb868c0, 0xc0001c8f20, 0x1b8, 0xb868c0, 0xc0001c89b0, 0x1b8, ...)
        /home/erhudy/go/src/bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff/diff_string.go:28 +0x11f
bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff.(*Changelog).diff(0xc0000cca60, 0xc0001c8fc0, 0x3, 0x3, 0xb868c0, 0xc0001c8f20, 0x1b8, 0xb868c0, 0xc0001c89b0, 0x1b8, ...)
        /home/erhudy/go/src/bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff/diff.go:90 +0xc96
bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff.(*Changelog).structValues(0xc00012fbe8, 0xc81af9, 0x6, 0xc0000bd300, 0x2, 0x2, 0xc22ac0, 0xc0001c8990, 0x199, 0x100, ...)
        /home/erhudy/go/src/bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff/diff_struct.go:86 +0x41c
bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff.(*Changelog).diffStruct(0xc0000cdbe8, 0xc0000bd300, 0x2, 0x2, 0x0, 0x0, 0x0, 0xc22ac0, 0xc0001c8990, 0x199, ...)
        /home/erhudy/go/src/bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff/diff_struct.go:18 +0x606
bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff.(*Changelog).diff(0xc0000cdbe8, 0xc0000bd300, 0x2, 0x2, 0x0, 0x0, 0x0, 0xc22ac0, 0xc0001c8990, 0x199, ...)
        /home/erhudy/go/src/bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff/diff.go:86 +0xdf6
bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff.(*Changelog).diffPtr(0xc0000cdbe8, 0xc0000bd300, 0x2, 0x2, 0x0, 0x0, 0x0, 0xc565c0, 0xc0001c8990, 0x16, ...)
        /home/erhudy/go/src/bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff/diff_pointer.go:15 +0x110
bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff.(*Changelog).diff(0xc0000cdbe8, 0xc0000bd300, 0x2, 0x2, 0x0, 0x0, 0x0, 0xc565c0, 0xc0001c8990, 0x16, ...)
        /home/erhudy/go/src/bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff/diff.go:102 +0x876
bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff.(*Changelog).diffComparative(0xc00012fbe8, 0xc0000a3f40, 0x1, 0x1, 0xc00012f3a8, 0xc0000b4f40, 0x2)
        /home/erhudy/go/src/bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff/diff_comparative.go:25 +0x34c
bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff.(*Changelog).diffMap(0xc0000cdbe8, 0xc0000a3f40, 0x1, 0x1, 0xbafea0, 0xc0000bd120, 0x195, 0xbafea0, 0xc0000bd140, 0x195, ...)
        /home/erhudy/go/src/bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff/diff_map.go:33 +0x579
bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff.(*Changelog).diff(0xc0000cdbe8, 0xc0000a3f40, 0x1, 0x1, 0xbafea0, 0xc0000bd120, 0x195, 0xbafea0, 0xc0000bd140, 0x195, ...)
        /home/erhudy/go/src/bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff/diff.go:100 +0x926
bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff.(*Changelog).diffStruct(0xc00012fbe8, 0x132dc18, 0x0, 0x0, 0xbf1c60, 0xc0000bd120, 0x199, 0xbf1c60, 0xc0000bd140, 0x199, ...)
        /home/erhudy/go/src/bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff/diff_struct.go:42 +0x44e
bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff.(*Changelog).diff(0xc0000cdbe8, 0x132dc18, 0x0, 0x0, 0xbf1c60, 0xc0000bd120, 0x199, 0xbf1c60, 0xc0000bd140, 0x199, ...)
        /home/erhudy/go/src/bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff/diff.go:86 +0xdf6
bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff.(*Changelog).diffPtr(0xc0000cdbe8, 0x132dc18, 0x0, 0x0, 0xc59420, 0xc0000bd120, 0x16, 0xc59420, 0xc0000bd140, 0x16, ...)
        /home/erhudy/go/src/bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff/diff_pointer.go:42 +0x506
bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff.(*Changelog).diff(0xc0000cdbe8, 0x132dc18, 0x0, 0x0, 0xc59420, 0xc0000bd120, 0x16, 0xc59420, 0xc0000bd140, 0x16, ...)
        /home/erhudy/go/src/bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff/diff.go:102 +0x876
bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff.Diff(0xc59420, 0xc0000bd120, 0xc59420, 0xc0000bd140, 0xc0000bd180, 0xc0000cdd48, 0xb10fd2, 0xb74a40, 0xc0000bd180)
        /home/erhudy/go/src/bbgithub.dev.bloomberg.com/babka/cli/vendor/github.com/r3labs/diff/diff.go:51 +0x1d5
[...my tests...]
testing.tRunner(0xc0000e0900, 0xca5a10)
        /home/erhudy/go1.12.5/src/testing/testing.go:865 +0xc0
created by testing.(*T).Run
        /home/erhudy/go1.12.5/src/testing/testing.go:916 +0x35a
FAIL    bbgithub.dev.bloomberg.com/babka/cli/pkg/babka/config   0.026s
```

I was able to work around this panic by modifying `diff_string.go` thusly:
```
@@ -22,7 +22,7 @@ func (cl *Changelog) diffString(path []string, a, b reflect.Value) error {
        }
 
        if a.String() != b.String() {
-               cl.add(UPDATE, path, a.Interface(), b.Interface())
+               cl.add(UPDATE, path, a.String(), b.String())
        }
 
        return nil
```

--- a/diff_pointer.go
+++ b/diff_pointer.go
@@ -4,10 +4,24 @@
 
 package diff
 
-import "reflect"
+import (
+	"reflect"
+)
 
 func (cl *Changelog) diffPtr(path []string, a, b reflect.Value) error {
 	if a.Kind() != b.Kind() {
+		if a.Kind() == reflect.Invalid {
+			if !b.IsNil() {
+				return cl.diff(path, reflect.ValueOf(nil), reflect.Indirect(b))
+			}
+		}
+
+		if b.Kind() == reflect.Invalid {
+			if !a.IsNil() {
+				return cl.diff(path, reflect.Indirect(a), reflect.ValueOf(nil))
+			}
+		}
+
 		return ErrTypeMismatch
 	}
 

--- a/diff_string.go
+++ b/diff_string.go
@@ -22,7 +22,7 @@ func (cl *Changelog) diffString(path []string, a, b reflect.Value) error {
 	}
 
 	if a.String() != b.String() {
-		cl.add(UPDATE, path, a.Interface(), b.Interface())
+		cl.add(UPDATE, path, a.String(), b.String())
 	}
 
 	return nil

--- a/diff_test.go
+++ b/diff_test.go
@@ -9,9 +9,13 @@ import (
 	"time"
 
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
 var currentTime = time.Now()
+
+var struct1 = tmstruct{Bar: 1, Foo: "one"}
+var struct2 = tmstruct{Bar: 2, Foo: "two"}
 
 type tistruct struct {
 	Name  string `diff:"name,identifier"`
@@ -308,16 +312,6 @@ func TestDiff(t *testing.T) {
 			ErrTypeMismatch,
 		},
 		{
-			"mismatched-values-struct-nil", nil, &tstruct{Identifiables: []tistruct{{"one", 1}}},
-			Changelog{},
-			ErrTypeMismatch,
-		},
-		{
-			"mismatched-values-nil-struct", &tstruct{Identifiables: []tistruct{{"one", 1}}}, nil,
-			Changelog{},
-			ErrTypeMismatch,
-		},
-		{
 			"omittable", tstruct{Ignored: false}, tstruct{Ignored: true},
 			Changelog{},
 			nil,
@@ -348,6 +342,26 @@ func TestDiff(t *testing.T) {
 			},
 			nil,
 		},
+		{
+			"map-string-pointer-create",
+			map[string]*tmstruct{"one": &struct1},
+			map[string]*tmstruct{"one": &struct1, "two": &struct2},
+			Changelog{
+				Change{Type: CREATE, Path: []string{"two", "foo"}, From: nil, To: "two"},
+				Change{Type: CREATE, Path: []string{"two", "bar"}, From: nil, To: 2},
+			},
+			nil,
+		},
+		{
+			"map-string-pointer-delete",
+			map[string]*tmstruct{"one": &struct1, "two": &struct2},
+			map[string]*tmstruct{"one": &struct1},
+			Changelog{
+				Change{Type: DELETE, Path: []string{"two", "foo"}, From: "two", To: nil},
+				Change{Type: DELETE, Path: []string{"two", "bar"}, From: 2, To: nil},
+			},
+			nil,
+		},
 	}
 
 	for _, tc := range cases {
@@ -355,7 +369,7 @@ func TestDiff(t *testing.T) {
 			cl, err := Diff(tc.A, tc.B)
 
 			assert.Equal(t, tc.Error, err)
-			assert.Equal(t, len(tc.Changelog), len(cl))
+			require.Equal(t, len(tc.Changelog), len(cl))
 
 			for i, c := range cl {
 				assert.Equal(t, tc.Changelog[i].Type, c.Type)

